### PR TITLE
New auth service and decorators

### PR DESCRIPTION
**Pull request by byewokko**
# Breaking changes
- `asab.web.authz` is deprecated and will be removed in January 2024. Use `asab.web.auth` instead.

# Migrating from  `asab.web.authz` in :six: easy steps
- Change `asab.web.authz.AuthzService` to `asab.web.auth.AuthService` and run its `install` method in init (see AuthService reference below).
- Change `@asab.web.authz.required()` to `@asab.web.auth.require()`.
- Delete `@asab.web.authz.required("tenant:access")`. Tenant access is checked implicitly by AuthService.
- Delete `@asab.web.tenant.tenant_handler` and  `@asab.web.authz.userinfo_handler`. The respective handler arguments are now automatically supplied by AuthService.
- Decorate the handlers that should not use authentication with `@asab.web.auth.noauth`.
- In the config file, create section `[auth]` and provide `public_keys_url` and `tenant_url`. Or enable dev mode (see below).

---

# Multitenancy

## Strictly multitenant endpoints

Strictly multitenant endpoints always operate within a tenant, hence they need the `tenant` parameter to be always provided. Such endpoints must define the `tenant` parameter in their **URL path** and include `tenant` argument in the handler method. Auth service extracts the tenant from the URL path, validates the tenant existence, checks if the request is authorized for the tenant and finally passes the tenant name to the handler method.

Example handler:
```python
class MenuHandler:
	def __init__(self, app):
		self.MenuService = app.get_service("MenuService")
		router = app.WebContainer.WebApp.router
		# Add a path with `tenant` parameter
		router.add_get("/{tenant}/todays-menu", self.get_todays_menu)

	# Define handler method with `tenant` argument in the signature
	async def get_todays_menu(self, request, *, tenant):
		menu = await self.MenuService.get_todays_menu(tenant)
		return asab.web.rest.json_response(request, data=menu)
```

Example request:
```
GET http://localhost:8080/lazy-raccoon-bistro/todays-menu
```

## Configurably multitenant endpoints

Configurably multitenant endpoints usually operate within a tenant, but they can also operate in tenantless mode if the application is configured so. When you create an endpoint *without* `tenant` parameter in the URL path and *with* `tenant` argument in the handler method, the Auth service will either expect the `tenant` parameter to be provided in the **URL query** if mutlitenancy is enabled, or to not be provided at all if multitenancy is disabled. Use the `[auth] multitenancy` boolean switch in config to control the multitenancy setting.

- If `multitenancy` is **enabled**, the request's **query string** must include a tenant parameter. Requests without the tenant query result in Bad request (HTTP 400).
- If `multitenancy` is **disabled**, the tenant argument in the handler method is set to `None`. Any `tenant` parameter in the query string is ignored.

Example handler:
```python
class MenuHandler:
	def __init__(self, app):
		self.MenuService = app.get_service("MenuService")
		router = app.WebContainer.WebApp.router
		# Add a path without `tenant` parameter
		router.add_get("/todays-menu", self.get_todays_menu)

	# Define handler method with `tenant` argument in the signature
	async def get_todays_menu(self, request, *, tenant):
		menu = await self.MenuService.get_todays_menu(tenant)
		return asab.web.rest.json_response(request, data=menu)
```

Example request (with `[auth] multitenancy=yes`):
```
GET http://localhost:8080/todays-menu?tenant=lazy-raccoon-bistro
```

Example request (with `[auth] multitenancy=no`):
```
GET http://localhost:8080/todays-menu
```

# Module reference

## `asab.web.auth.AuthService`

Provides communication with authorization server and authenticates and authorizes requests.

Usage:
```python
class MyApplication(asab.Application):
	def __init__(self):
		super().__init__()
		self.add_module(asab.web.Module)
		self.WebService = self.get_service("asab.WebService")
		self.WebContainer = asab.web.WebContainer(self.WebService, "web")

		# Initialize authorization service and install the decorators
		self.AuthzService = asab.web.auth.AuthService(self)
		self.AuthzService.install(self.WebContainer)
```

Installing the service allows you to use special keyword args in your handlers (`tenant`, `user_info` and `resources`) that will be automatically filled in by the service, based on the request and its authorization.

Usage:
```python
async def order_breakfast(self, request, *, tenant: str, user_info: dict, resources: frozenset):
	who = user_info["sub"]
	if "full-english:eat" in resources or "pancakes:eat" in resources:
		...
	return asab.web.rest.json_response(request, {"result": "OK"})
```

### Service configuration

In minimum **production** config, you need to specify at least `public_keys_url` and `tenant_url`, for example:
```ini
[auth]
public_keys_url=http://localhost:8081/openidconnect/public_keys
tenant_url=http://localhost:8081/tenant
```

In minimum **development** config, you can just activate the **dev mode**, in which the service does not need the above URLs configured:
```ini
[auth]
dev_mode=yes
```

## `asab.web.auth.require()` decorator

Specify resources that are required to access the endpoint. Failure results in HTTP 403.

Usage:
```python
@asab.web.auth.require("cake:access", "cake:eat")
async def eat_cake(self, request, *, user_info):
	await CafeService.eat("cake", user=user_info["sub"])
	return asab.web.rest.json_response(request, {"result": "OK"})
```

## `asab.web.auth.noauth` decorator

Skips request authentication and authorization for this endpoint. No user info/tenants/resources are available in the handler method.

Usage:
```python
@asab.web.auth.noauth
async def drink_soda(self, request):
	await CafeService.drink("soda")
	return asab.web.rest.json_response(request, {"result": "OK"})
```

# Development

## Dev mode
Disables actual autentication and authorization. Useful for developing ASAB apps without authorization server. Activate by enabling `[auth] dev_mode`. You can also specify a path to a JSON file with your own mocked userinfo payload `[auth] dev_user_info_path`:

```ini
[auth]
dev_mode=yes
dev_user_info_path=${THIS_DIR}/mock_userinfo.json
```

When dev mode is enabled, you don't have to provide `public_keys_url` and `tenant_url` as these options are ignored.

## Usage

See `examples/web-authz.py` for usage example.

### CODE REVIEW

**Review by ateska — CHANGES_REQUESTED** [review]

File: asab/web/auth/decorator.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+				raise aiohttp.web.HTTPForbidden()
+
+			return await handler(*args, **kwargs)
+
+		return wrapper
+
+	return decorator_required
+
+
+def require_superuser(handler):
```
[the remark this answers is not in the document]
Comment (byewokko) [reply asab/web/auth/decorator.py]: removed

File: asab/web/auth/__init__.py [hunk 2]
```diff
@@ -0,0 +1,8 @@
+from .decorator import require, no_auth
+from .service import AuthService
+
+__all__ = (
+	"AuthService",
+	"require",
+	"no_auth",
```
[the remark this answers is not in the document]
Comment (byewokko) [reply asab/web/auth/__init__.py]: renamed

**Comment by eliska-n** [comment]
bsquery tenant-aware option without resources
dev mode seems not to work...
or do I need to mock user info?

```
[auth]
dev_mode=yes
```

```
class BSQueryApp(asab.Application):

	def __init__(self):
		super().__init__()

		self.add_module(asab.web.Module)
		self.add_module(asab.zookeeper.Module)
		self.add_module(asab.metrics.Module)

		self.ASABApiService = asab.api.ApiService(self)

		# Initialize web container
		self.add_module(asab.web.Module)
		self.WebService = self.get_service("asab.WebService")
		self.WebContainer = asab.web.WebContainer(self.WebService, "web")

		# Initialize authorization
		self.AuthService = asab.web.auth.AuthService(self)
		self.WebContainer.WebApp.middlewares.append(asab.web.rest.JsonExceptionMiddleware)
		self.AuthService.install(self.WebContainer)
		...
```

asab.web.auth.service No Authorization header.

File: asab/web/auth/service.py [hunk 3]
```diff
@@ -0,0 +1,435 @@
...
+		"public_keys_url": "",
+
+		# Whether the app is tenant-aware
+		"multitenancy": "yes",
+
+		# In DEV MODE
+		# - no authorization server is needed,
+		# - all incoming requests are "authorized" with custom mocked user info data loaded from a JSON file,
+		"dev_mode": "no",
+		"dev_user_info_path": "",
```
Comment (byewokko) [inline asab/web/auth/service.py:74]: @eliska-n (I can't reply directly to your comment)
Yes, the dev mode works by injecting mock userinfo to every request. If you want to access everything, you need to overwrite the default userinfo by providing your own superuser userinfo file. 
You specify `dev_user_info_path=/path/to/userinfo.json` and inside that file just put
```json
{
 "resources": {
  "*": ["authz:superuser"],
  "my-tenant": ["authz:superuser"],
  "another-tenant": ["authz:superuser"],
 }
}
```

Comment (byewokko) [reply asab/web/auth/service.py:74]: Now that I think about it, I'll add the superuser resource to the default userinfo. That will be the most common usage anyways.

**Pull request merged**